drivers/input/touchscreen/himax.c

```diff
@@ -0,0 +1,462 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Driver for Himax touchscreens
+ *
+ * Copyright (C) 2022 Job Noorman <job@noorman.info>
+ *
+ * This code is based on "Himax Android Driver Sample Code for QCT platform":
+ *
+ * Copyright (C) 2017 Himax Corporation.
+ */
+
+#include <linux/kernel.h>
+#include <linux/err.h>
+#include <linux/delay.h>
+#include <linux/interrupt.h>
+#include <linux/i2c.h>
+#include <linux/gpio/consumer.h>
+#include <linux/input.h>
+#include <linux/input/mt.h>
+#include <linux/input/touchscreen.h>
+#include <asm/byteorder.h>
+
+#define HIMAX_GPIO_RST_NAME "reset"
+
+#define HIMAX_ID_83112A 0x83112a
+#define HIMAX_ID_83112B 0x83112b
+
+#define HIMAX_MAX_POINTS 10
+
+#define HIMAX_CMD_REG_SET_ADDR 0x00
+#define HIMAX_CMD_REG_INIT_READ 0x0c
+#define HIMAX_CMD_REG_READ_VALUE 0x08
+#define HIMAX_CMD_READ_EVENT 0x30
+
+#define HIMAX_REG_PRODUCT_ID 0x900000d0
+
+struct himax_event_point {
+	__be16 x;
+	__be16 y;
+} __packed;
+
+struct himax_event {
+	struct himax_event_point points[HIMAX_MAX_POINTS];
+	u8 majors[HIMAX_MAX_POINTS];
+	u8 pad0[2];
+	u8 num_points;
+	u8 pad1[2];
+	u8 checksum_fix;
+} __packed;
+
+static_assert(sizeof(struct himax_event) == 56);
+
+struct himax_ts_data {
+	struct gpio_desc *gpiod_rst;
+	struct input_dev *input_dev;
+	struct i2c_client *client;
+	struct touchscreen_properties props;
+};
+
+static int himax_i2c_read(struct himax_ts_data *ts, u8 command, u8 *data,
+			  u8 length)
+{
+	int ret;
+	struct i2c_client *client = ts->client;
+
+	// clang-format off
+	struct i2c_msg msg[] = {
+		{
+			.addr = client->addr,
+			.flags = 0,
+			.len = 1,
+			.buf = &command,
+		},
+		{
+			.addr = client->addr,
+			.flags = I2C_M_RD,
+			.len = length,
+			.buf = data,
+		}
+	};
+	// clang-format on
+
+	ret = i2c_transfer(client->adapter, msg, ARRAY_SIZE(msg));
```

Comment:
Looks very interesting. I'll take a look asap!

Reply:
Done! `regmap` is very interesting, thanks for pointing this out! Also, another 66 LoC down the drain :)

The only thing I find a bit dodgy is that I have to use `regmap_raw_read` [here](https://github.com/msm8953-mainline/linux/pull/43/files#diff-04d74c15495a45ed581f601eb480981f6bf3a82cd0cc1046c20f8f71121b1a85R84-R85) to avoid endianness conversion. Also, we are kind of lucky that `sizeof(struct himax_event)` is a multiple of `himax_regmap_config.val_bits / 8` because otherwise it wouldn't have been possible to use `regmap` here at all afaict.

Reply:
I'd imagine that there must be something so this works correctly? Some flag set somewhere?

Reply:
The problem is the accessing the "normal" registers (everything except the event input) *does* need endianness conversion.

But I guess it's fine to use `regmap_raw_read` as its documentation says "Read raw data from the device" which fits the event struct imo.

(btw, everything works correctly, I just found the use of `regmap_raw_read` a bit weird because no other touchscreen drivers use it. There are a lot of other drivers using it though.)